src/main/java/com/joejoe2/chat/models/GroupMessage.java

```diff
@@ -79,6 +82,30 @@ public static GroupMessage leaveMessage(GroupChannel channel, User actor, User s
     return message;
   }
 
+  /** generate a message for admin banning an user in the GroupChannel */
+  public static GroupMessage banMessage(GroupChannel channel, User actor, User subject) {
+    GroupMessage message = new GroupMessage();
+    message.channel = channel;
+    message.from = actor;
+    message.content =
+        "{\"id\":\"%s\", \"username\":\"%s\"}"
+            .formatted(subject.getId().toString(), subject.getUserName());
+    message.messageType = MessageType.BAN;
+    return message;
+  }
+
+  /** generate a message for admin unbanning an user in the GroupChannel */
+  public static GroupMessage unbanMessage(GroupChannel channel, User actor, User subject) {
```

Comment:
because their format at almost same. I would like to extract the same part like:
```
private static GroupMessage generateBanOrUnBanMessage(GroupChannel channel, User actor, User subject){
    GroupMessage message = new GroupMessage();
    message.channel = channel;
    message.from = actor;
    message.content =
            "{\"id\":\"%s\", \"username\":\"%s\"}"
                    .formatted(subject.getId().toString(), subject.getUserName());
  }

  /** generate a message for admin banning an user in the GroupChannel */
  public static GroupMessage banMessage(GroupChannel channel, User actor, User subject) {
    GroupMessage message = generateBanOrUnBanMessage(channel, actor, subject);
    message.messageType = MessageType.BAN;
    return message;
  }

  /** generate a message for admin unbanning an user in the GroupChannel */
  public static GroupMessage unbanMessage(GroupChannel channel, User actor, User subject) {
    GroupMessage message = generateBanOrUnBanMessage(channel, actor, subject);
    message.messageType = MessageType.UNBAN;
    return message;
  }
```

Reply:
In general, it's advised to have a single method which creates a proper format and reuse for both ban/unban occurrences.

Reply:
If we have genrateBanMessage(GroupChannel channel, User actor, User subject, flag), how to phrase the 'flag'? I cannot come up with a reasonable name.

Reply:
Instead of a flag, we can pass the _MessageType_ enum in the place of use

Reply:
and you call name it just _messageType_

Reply:
I think using messageType here is not suitable because there are other types we need to check. Another sol:
``` 
private static GroupMessage banOrUnBanMessageFormat(GroupChannel channel, User actor, User subject){
    GroupMessage message = new GroupMessage();
    message.channel = channel;
    message.from = actor;
    message.content =
            "{\"id\":\"%s\", \"username\":\"%s\"}"
                    .formatted(subject.getId().toString(), subject.getUserName());
     return message;
  }
```
and we call this in the group channel and manually set the messageType.

Reply:
Or warp like a builder:
```
public static class BanOrUnBanMessageBuilder {
    private GroupMessage message;
    public BanOrUnBanMessageBuilder(GroupChannel channel, User actor, User subject){
      message = new GroupMessage();
      message.channel = channel;
      message.from = actor;
      message.content =
              "{\"id\":\"%s\", \"username\":\"%s\"}"
                      .formatted(subject.getId().toString(), subject.getUserName());
    }

    public GroupMessage ban(){
      message.messageType = MessageType.BAN;
      return message;
    }

    public GroupMessage unban(){
      message.messageType = MessageType.UNBAN;
      return message;
    }
  }

  private static GroupMessage banOrUnBanMessageFormat(
      GroupChannel channel, User actor, User subject) {
    GroupMessage message = new GroupMessage();
    message.channel = channel;
    message.from = actor;
    message.content =
        "{\"id\":\"%s\", \"username\":\"%s\"}"
            .formatted(subject.getId().toString(), subject.getUserName());
    return message;
  }
```

Reply:
In this case there is more copy-paste. The method with paramter is a more favourable approach which is usually used.

Reply:
If we have an enum called _ChatActions_ as mentioned in the latest review then this can be one of the params and the param itself might be called - _action_

Reply:
I think our discussions made this things a little complicated 🤔. I would prefer:
1. remain current two methods and extract the same part.
2. use one method with a flag isBan and document it explicitly, but i think 1. will be more readable

Reply:
Sure, both options work just fine, pick up one that you prefer the most :)